### ISSUE TITLE
Seeing "Could Not Delete File" from MultiPartFormInputStream after migrating from Jetty 9 to Jetty 11

### ISSUE THREAD

**Issue by tipsy** [eclipse/jetty.project#8417]
**Jetty version(s)**
11.0.11

**Java version/vendor** `(use: java -version)`
openjdk-18.0.1.1

**OS type/version**
Windows 10

**Description**

```java
[JettyServerThreadPool-25] WARN org.eclipse.jetty.server.Request - Errors deleting multipart tmp files
java.lang.RuntimeException: java.io.IOException: Could Not Delete File
	at org.eclipse.jetty.util.MultiException.ifExceptionThrowRuntime(MultiException.java:148)
	at org.eclipse.jetty.server.MultiPartFormInputStream.delete(MultiPartFormInputStream.java:463)
	at org.eclipse.jetty.server.MultiPartFormInputStream.deleteParts(MultiPartFormInputStream.java:438)
	at org.eclipse.jetty.server.MultiParts$MultiPartsHttpParser.close(MultiParts.java:101)
	at org.eclipse.jetty.server.Request.onCompleted(Request.java:1463)
	at org.eclipse.jetty.server.HttpChannel.onCompleted(HttpChannel.java:966)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:485)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:282)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:319)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:894)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1038)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.io.IOException: Could Not Delete File
	at org.eclipse.jetty.server.MultiPartFormInputStream$MultiPart.delete(MultiPartFormInputStream.java:328)
	at org.eclipse.jetty.server.MultiPartFormInputStream$MultiPart.cleanUp(MultiPartFormInputStream.java:339)
	at org.eclipse.jetty.server.MultiPartFormInputStream.delete(MultiPartFormInputStream.java:450)
	... 12 more
```

**How to reproduce?**
Creating a minimal broken example would be a bit of effort, but I can have a look at doing that later if it's required. The code I'm running can be found here: https://github.com/javalin/javalin/blob/master/javalin/src/main/java/io/javalin/http/util/MultipartUtil.kt

## TL;DR: - Jetty logging changed, I had a bug in my code

> @janbartel I see the difference between jetty-9.4 and jetty-11: in jetty-9.4 we just call file.delete(), which does not throw an exception on failure, whereas in jetty-11, we check the result of file.delete() and throw an exception on failure to delete it. So that explains why an exception in jetty-11 and not in jetty-9.4. ... Do you call Part.getInputStream() in your code? If so, do you ensure that you call close() on it?

(from https://github.com/eclipse/jetty.project/issues/8417#issuecomment-1207511370)

**Comment by joakime** [comment]
This is typical Windows file locking behaviors.

It's nearly impossible to quickly delete or atomically delete a file that was recently used on windows from Java.

We have no workaround, this is bug that exists between Windows and Java with no good workaround.

**Comment by tipsy** [comment]
Let me try it on my mac to confirm :)

**Comment by tipsy** [comment]
Can confirm that it works as before on MacOS. I find it strange that there is a change in behavior from Jetty 9 to Jetty 11 (I went back to Jetty 9 to check).

Is it possible to configure either the behavior (automatically deleting) or the logging? This being a `log.warn` will definitely scare people who see it in development, even if it won't be an issue in production.

**Comment by joakime** [comment]
IIRC Jetty 9 used `File.deleteOnExit()` in this code path.
Which would only delete the file on exit of the JVM.
That's why it never seems to fail on Jetty 9, it doesn't actually delete the file until the JVM exits.
As you probably guessed, that's a horrible API to use on a server.

**Comment by tipsy** [comment]
I see. I think the current behavior is a little problematic. I'll use Javalin as an example, but I think this also holds true for the bigger JVM frameworks which rely on Jetty: The majority of our app developers sit on Windows machines (~37% based on statistics from our documentation website). These developers are going to see and worry about these warnings during development. I don't have statistics for my next point, but I believe the vast majority of these developers do not run their app on Windows in production. This means that all of these developers will see this (rather scary) warning when they are working on their apps, but they have no hope of fixing it, and they don't actually need to fix anything.

Some ideas for how this can be improved:

1. Keep a queue of files to delete, and delete them when Windows releases them
2. Revert to using `File.deleteOnExit()` for Windows (and log a warning about this on server startup?)

I would of course prefer 1, but 2 is probably better considering how few people actually run on Windows. It wouldn't even be a breaking change, since this is what Jetty is effectively doing now (it's not able to delete the files anyway).

**Comment by joakime** [comment]
> 1. Keep a queue of files to delete, and delete them when Windows releases them

How do you detect, from Java, that windows releases them?

> 2. Revert to using `File.deleteOnExit()` for Windows (and log a warning about this on server startup?)

On heavy Jetty users, with MultiPart, this leads to a OutOfMemory error.
This will reintroduce a terrible bug, and is not a valid option.

**Comment by tipsy** [comment]
> On heavy Jetty users, with MultiPart, this leads to a OutOfMemory error.
> This will reintroduce a terrible bug, and is not a valid option.

How are these users dealing with files not being deleted in the current implementation then?

> How do you detect, from Java, that windows releases them?

I am completely ignorant on this topic, but I imagine you could have some retry interval and retry limit.

**Comment by joakime** [comment]
> > On heavy Jetty users, with MultiPart, this leads to a OutOfMemory error.
> > This will reintroduce a terrible bug, and is not a valid option.
> 
> How are these users dealing with files not being deleted in the current implementation then?

On Windows, its a similar problem.
The file is not even attempted to be deleted on Jetty 9 until the JVM exits.
The file is not deleted in many cases (there are cases with specific FileSystem setups where they will delete) on Jetty 10+.

So, if you are Windows user you have a resource leak on Jetty 9+

> > How do you detect, from Java, that windows releases them?
> 
> I am completely ignorant on this topic, but I imagine you could have some retry interval and retry limit.

Thing is, on some Windows FileSystems (or other configurations impacting the FileSystem) the file is never released until the JVM exits all the way.
Even the `File.deleteOnExit()` solution from Jetty 9 suffers from this bug in Windows + Java (no delete occurs).

Be aware that this issue is not a 100% guaranteed behavior on Windows.

We've had reports that the Microsoft's Azure Windows servers don't suffer from this file locking behavior. (don't know what's special about those environments)

Some far more intelligent folks have tried to figure this out, and so far the fingers seem to point to security software (virus scanners and the like).

We've tried a variety of things to solve this, but we keep hitting the "until the JVM exits" problem.
Example: A periodic sweep of files that are past some kind of timeout against lastModified (still couldn't delete due to Lock held by Windows).

**Comment by joakime** [comment]
Here's an example of a past experiment on this ..

https://github.com/eclipse/jetty.project/pull/6588/files

We tried using the `java.lang.ref.Cleaner` to track when a reference to the file and it's resources are completely removed from Java.
Leaving only the Windows file locks left to deal with.
Still didn't work.
Why?  The locks remained until the JVM exited.
Hard to write something in Java that needs to execute once the JVM has exited.

**Comment by dzikoysk** [comment]
Hmm... In general this issue is caused by using [`FileChannel#map`](https://docs.oracle.com/javase/8/docs/api/java/nio/channels/FileChannel.html#map-java.nio.channels.FileChannel.MapMode-long-long-) (creating an instance of MappedByteBuffer/DirectByteBuffer) at some point, because using standard I/O on channel should work just fine? 🤔 Is there any chance to not call this method at all, through some kind of OS-dependent impl. It's more likely just an optimization, but as long as it's not even properly handled by JVM... 🤷 

The only explicit call to this method I can find is here, so this is the source of this issue I guess?
https://github.com/eclipse/jetty.project/blob/5e10334cbb77825ad8d4f17d066fe424613a1960/jetty-server/src/main/java/org/eclipse/jetty/server/handler/FileBufferedResponseHandler.java#L211

**Comment by janbartel** [comment]
@tipsy can you enable debug in jetty-9.4? IIRC we register a listener that will clean up temporary multipart files when the Request exits the container (see https://www.eclipse.org/jetty/javadoc/jetty-9/org/eclipse/jetty/server/MultiPartCleanerListener.html), not just delete them when the jvm exits. If we do have that listener, then it is strange that 9.4 is successfully deleting the files, whereas jetty-11 is not. 

@dzikoysk depends how the the part is used: for example, if the user is calling eg `Part.getInputStream()` then we return `BufferedInputStream(new FileInputStream(_file))` if the part was written to a file, or otherwise we return a `ByteArrayInputStream` with the in-memory bytes.

**Comment by dzikoysk** [comment]
Hmm I see, maybe this behavior could be moved to some kind of configurable factory? E.g. `Part#getInputStream` expects any kind of `InputStream` implementation, not `BufferedInputStream` explicitly, so it could be up to the factory to respond with given implementation tho 🤔 Buffering is pretty much about speeding up IO, so as long as we wouldn't care that much about it in our scenario, we could just provide standard file stream on Windows.

**Comment by tipsy** [comment]
> @janbartel @tipsy can you enable debug in jetty-9.4? 

Sure !

9.4: https://pastebin.com/raw/pBg4P01N
11.0.11: https://pastebin.com/raw/EG66QYkd

It's quite a change in level of output between the two versions.

**Comment by janbartel** [comment]
@tipsy thanks for that. I see the difference between jetty-9.4 and jetty-11:  in jetty-9.4 we just call `file.delete()`, which does not throw an exception on failure, whereas in jetty-11, we check the result of `file.delete()` and throw an exception on failure to delete it.  So that explains why an exception in jetty-11 and not in jetty-9.4.

As to why those temporary files are locked on windows, I don't have enough information to say. Do you call `Part.getInputStream()` in your code? If so, do you ensure that you call `close()` on it? Do you interact in any other way with any of the multiparts?

**Comment by tipsy** [comment]
> So that explains why an exception in jetty-11 and not in jetty-9.4.

Thank you very much!

> Do you call Part.getInputStream() in your code? If so, do you ensure that you call close() on it?

![image](https://user-images.githubusercontent.com/1521451/183375249-76c4298b-0716-4db1-8407-7e0bcbab78ec.png)

---

I am able to resolve this in my project now.

Thanks a lot for the support, and feel free to close this issue (I don't trust myself to do it).

**Comment by olamy** [comment]
seems to be fixed ;)
